android/titanium/src/java/org/appcelerator/titanium/TiApplication.java

```diff
@@ -769,40 +796,68 @@ public boolean isCoverageEnabled()
 		return false;
 	}
 
+	/**
+	 * Determines if main Titanium script has been loaded via launch() method and is currently running.
+	 * @return
+	 * Returns true if main Titanium script has been launched and is actively running.
+	 * <p>
+	 * Returns false if script has never been loaded or has runtime has been disposed/terminated.
+	 */
+	public static boolean isScriptRunning()
+	{
+		return tiApp.isScriptRunning;
+	}
+
 	public static void launch()
 	{
+		// Do not continue if script has already been loaded.
+		if (tiApp.isScriptRunning) {
+			return;
+		}
+
+		// Fetch the root activity that which will host the runtime.
+		ActivityProxy activityProxy = null;
 		final TiRootActivity rootActivity = TiApplication.getInstance().getRootActivity();
-		if (rootActivity == null) {
+		if (rootActivity != null) {
+			activityProxy = rootActivity.getActivityProxy();
+		}
+
+		// If runtime's lifetime is bound to the UI, then do not continue if failed to find root activity.
+		// We only allow this if "run-in-background" setting is enabled in "tiapp.xml".
+		if (KrollRuntime.isBoundToUI() && (rootActivity == null)) {
```

Comment:
Looks like you bail before loading the snapshot/ti.main.js here if the app is "running on background". The issue is that may cause some confusion from devs who may expect polyfills/node shims/whatever. I assume we'd have to modify our `common` js code to be more aware of whether there's a UI session to avoid doing some of it's work when in the BG?

Reply:
This if-condition maintains our old behavior when backgrounding is disabled where we only support loading JS files once the UI is available. The root activity is responsible for loading "ti.main" and guarantees that our extensions, polyfills, etc. are available. So, we're okay here. (I'll double check our JS-Service handling, which had broken handing before but works fine with our new backgrounding feature.)

When backgrounding is enabled, we load "ti.main" during `Application.onCreate()` which gets called before any activities are created and leave the JS runtime running for the duration of the app process.